docs/how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,124 @@
+library encountered. Whether or not the build tool produces a graph as an output
+of this resolution, it is useful to think of this output in terms of a new
+graph. We call the graph representation of this output a *resolved dependency
+graph*. Here is a visual representation of the process:
+
+<img src="assets/images/ddc-resolution-00.png">
+
+For example, we could have an unresolved graph that looks like this:
+
+<img src="assets/images/ddc-resolution-01.png">
+
+In this graph, there are two versions of A (10.0 and 10.1). The version
+resolution process picks a single version for A, which could be either version
+depending on the algorithm. If A:10.0 is picked, the resolved dependency graph
+looks like the following. In this graph, C's dependency on A:10.1 is overridden.
+
+<img src="assets/images/ddc-resolution-02.png">
+
+If A:10.1 is picked, the resolved dependency graph looks like the following. In
+this graph, B's dependency on A:10.0 is overridden.
+
+<img src="assets/images/ddc-resolution-03.png">
+
+So why do we need to pick a single version? The reason is that Java's default
+class loader only loads one version of each fully-qualified class name at
+runtime. So, if you try to put multiple versions on the classpath, you would not
+get both versions at runtime. Java build systems handle this issue upfront by
+picking a single version for each library before the classpath is even
+constructed.
+
+Now let's look at the version resolution algorithms used by Gradle and
+Maven. Let's consider Gradle first because it is the simpler one to
+describe. Simply put, Gradle chooses the highest version encountered in your
+dependency graph. For the example above, Gradle would pick A:10.1.
+
+<img src="assets/images/ddc-resolution-04.png">
+
+Gradle applies the higher-version rule even if it overrides one of your own
+direct dependency declarations:
+
+<img src="assets/images/ddc-resolution-05.png">
+
+Next let's consider Maven. Maven chooses the version that is closest to the root
+of the dependency graph, and if there is a tie, then it chooses the version it
+encounters first. Another way to describe the algorithm is that Maven chooses
+the first version encountered when traversing the dependency graph following a
+breadth-first traversal. Looking at the same example we used first for Gradle:
+
+<img src="assets/images/ddc-resolution-06.png">
+
+As you can see if you compare the resolved graphs, Maven makes the opposite
+choice. Let's look at the second example, where the root has both a direct and
+indirect dependency on the same library:
+
+<img src="assets/images/ddc-resolution-07.png">
+
+In this case, Maven chooses A:10.0 because it is only 1 hop away from the root
+instead of 2 hops (A:10.1). In this case, Maven also makes the opposite choice
+of Gradle.
+
+## Interaction between Maven and Gradle
+
+Every time a library is built, its build system performs version resolution for
```

Comment:
Every time --> When
its build system --> the build system
(some libraries are built with different systems; e.g. blaze)

Reply:
"the build system" also implies there is only one... I don't feel like it clarifies anything. Going into the nuances of multiple build systems per library seems like it would muddy the picture too much and make the article less useful. In the end, only 1 build system is used to build the artifacts that are pushed to Maven, even if a second build system is available to use.